### ISSUE TITLE
RFC: align GRPO/Distillation multi-validation with DPO's dict-of-dataloaders pattern

### ISSUE THREAD

**Issue by bzantium** [NVIDIA-NeMo/RL#2500]
This is a design discussion, not a bug. Filing as an RFC so the API direction can be settled before code review. Follows up #2497 / PR #2499 which shipped Option A (a non-breaking quick fix); this issue is Option B (the proper alignment).

## Background

Three of the four algorithms in NeMo-RL run validation:

| Algorithm | val data flow | Per-dataset metrics today |
|---|---|---|
| GRPO | single concatenated `AllTaskProcessedDataset` -> single `StatefulDataLoader` | No (PR #2499 added `accuracy_<task>` key suffixes as a workaround) |
| Distillation | same | same |
| DPO | `dict[str, AllTaskProcessedDataset]` -> `dict[str, StatefulDataLoader]` -> per-dataset loop with `validation-{name}/...` prefix | Yes, naturally |

DPO's design at `nemo_rl/algorithms/dpo.py:107-219, 332-377` is the cleaner pattern: per-dataset dataloaders, per-dataset wandb prefix, no slug-suffix in metric names, and trivially supports per-dataset cadence/batch-size knobs in the future.

GRPO and Distillation route through `nemo_rl/data/utils.py:setup_response_data:155-211`, which concatenates all val datasets into one Hugging Face `Dataset` before the dataloader is built:

```python
# nemo_rl/data/utils.py:200-211
val_dataset = None
if len(val_data_list) > 0:
    merged_val_data = concatenate_datasets(val_data_list)
    val_dataset = AllTaskProcessedDataset(
        merged_val_data,
        tokenizer,
        None,
        val_task_data_processors,
        ...
    )
```

The per-sample `task_name` survives in the batch (which is what PR #2499 leverages), but the dataloader-level identity is gone. That is why GRPO/Distillation needed a per-key suffix workaround in PR #2499 instead of the prefix-grouping that DPO gets for free.

## Proposal

Restructure the response-dataset path so it stays as a `dict[str, ...]` end to end, then update GRPO and Distillation `validate()` to mirror DPO. Concretely:

### `nemo_rl/data/utils.py:setup_response_data`

Return `dict[str, AllTaskProcessedDataset]` instead of a single concatenated dataset. Preserve dataset names as keys.

```python
# was: merged_val_data = concatenate_datasets(val_data_list); val_dataset = AllTaskProcessedDataset(merged_val_data, ...)
val_dataset = {
    name: AllTaskProcessedDataset(ds, tokenizer, None, val_task_data_processors, ...)
    for name, ds in zip(val_dataset_names, val_data_list)
}
```

### `nemo_rl/algorithms/grpo.py` and `distillation.py` setup

Build `val_dataloader: dict[str, StatefulDataLoader]` (mirrors `dpo.py:205-219`):

```python
val_dataloader = {
    name: StatefulDataLoader(ds, batch_size=val_batch_size, shuffle=False, collate_fn=..., drop_last=False)
    for name, ds in val_dataset.items()
}
```

### `validate()` in both files

Mirror `dpo.validate` (`dpo.py:332-377`): outer loop over `val_dataloader.items()`, factor the inner loop into a `validate_one_dataset(name, dataloader, ...)` helper, and `logger.log_metrics(metrics, step, prefix=f\"validation-{name}\")`.

Result in wandb: `validation-gsm8k/accuracy`, `validation-data-math500/accuracy`, `timing/validation-gsm8k/total_validation_time`, etc.

### Entry scripts

`examples/run_grpo.py`, `examples/run_distillation.py` need to thread the new `dict` return type through. Should be mechanical (the dict is just passed straight to setup).

## Why bother (the trade-off vs Option A)

| | Option A (PR #2499) | Option B (this RFC) |
|---|---|---|
| User-visible per-task metrics | Yes (`validation/accuracy_<task>`) | Yes (`validation-<task>/accuracy`) |
| Architectural consistency with DPO | Asymmetric (3 algorithms, 2 patterns) | Aligned |
| Slug-in-metric-name awkwardness | Yes for hyphenated names (`accuracy_data-math500`) | None |
| Per-dataset val cadence (future) | Hard to retrofit | Trivial |
| `setup_response_data` touched | No | Yes |
| Caller surface change | None | `Optional[AllTaskProcessedDataset]` -> `Optional[dict[str, AllTaskProcessedDataset]]` |
| Backwards compat | Full (legacy `validation/accuracy` preserved) | Breaking — see below |

## Migration concerns

1. **Legacy `validation/accuracy` key.** PR #2499 keeps it as the sample-weighted mean across all datasets. Option B removes it in favour of `validation-<name>/accuracy`. Two options to discuss:
   - Hard removal with a release-note migration guide.
   - One-version compatibility shim that also emits the legacy aggregated key, marked deprecated.
2. **`max_val_samples` semantics.** Today: cap on the concatenated stream. Under Option B: per-dataset cap, or global cap that round-robins, or the field becomes a `dict[str, int]`. Each has trade-offs; needs a decision.
3. **`val_batch_size`.** Same question — single value applied to all dataloaders, or `dict[str, int]`?
4. **`setup_response_data` return type.** Other callers of this function (if any) need to be audited.

## Open questions for maintainers

1. Is the architectural alignment worth the breaking change to the `validation/accuracy` key, or should the ecosystem stay on the suffix workaround?
2. If yes, should there be a one-release deprecation window with both keys emitted?
3. Should `max_val_samples` and `val_batch_size` also become per-dataset (`dict[str, int]`), or stay scalar?
4. Naming: `validation-{name}` follows DPO; alternative could be `validation/{name}/...` (slash-separated). Any preference?

## Out of scope

- DPO is already correct; no change needed there.
- The NeMo-Gym path in GRPO (`examples/nemo_gym/run_grpo_nemo_gym.py:173-187`) clobbers `val_batch_size` to `len(val_dataset)`; how this interacts with a dict-of-dataloaders setup needs its own thread.

## Prerequisite

PR #2499 (Option A) should land first so users have working per-task metrics in the meantime. Option B can then either supersede it cleanly or sit on top of it during a deprecation window.

Happy to draft the implementation PR once direction is settled. Tagging as design discussion rather than picking up the work unilaterally because the breaking change to `validation/accuracy` deserves maintainer sign-off.
